txnkv/transaction/2pc.go

```diff
@@ -1268,19 +1346,105 @@ func keepAlive(
 					}
 					return
 				}
-				continue
+			} else {
+				keepFail = 0
+				metrics.TxnHeartBeatHistogramOK.Observe(time.Since(startTime).Seconds())
+			}
+
+			// broadcast to all stores
+			if isPipelinedTxn {
+				broadcastToAllStores(
+					c.store,
+					retry.NewBackofferWithVars(
+						context.Background(),
+						broadcastMaxBackoff,
+						c.txn.vars,
+					),
+					&kvrpcpb.TxnStatus{
+						StartTs:     c.startTS,
+						MinCommitTs: c.minCommitTS.get(),
+						CommitTs:    0,
+						RolledBack:  false,
+						IsCompleted: false,
+					},
+					c.resourceGroupName,
+					c.resourceGroupTag,
+				)
 			}
-			keepFail = 0
-			metrics.TxnHeartBeatHistogramOK.Observe(time.Since(startTime).Seconds())
 		}
 	}
 }
 
-func sendTxnHeartBeat(bo *retry.Backoffer, store kvstore, primary []byte, startTS, ttl uint64) (newTTL uint64, stopHeartBeat bool, err error) {
+const broadcastRpcTimeout = time.Millisecond * 500
+const broadcastMaxConcurrency = 10
+
+// broadcasts to all stores to update the minCommitTS. Ignore errors.
+func broadcastToAllStores(
+	store kvstore,
+	bo *retry.Backoffer,
+	status *kvrpcpb.TxnStatus,
+	resourceGroupName string,
+	resourceGroupTag []byte,
+) {
+	stores := store.GetRegionCache().GetStoresByType(tikvrpc.TiKV)
+	req := tikvrpc.NewRequest(
+		tikvrpc.CmdBroadcastTxnStatus, &kvrpcpb.BroadcastTxnStatusRequest{
+			TxnStatus: []*kvrpcpb.TxnStatus{status},
+		},
+	)
+	req.Context.ClusterId = store.GetClusterID()
+	req.Context.ResourceControlContext = &kvrpcpb.ResourceControlContext{
+		ResourceGroupName: resourceGroupName,
+	}
+	req.Context.ResourceGroupTag = resourceGroupTag
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(stores))
+	semaphore := make(chan struct{}, broadcastMaxConcurrency)
+
+	for _, s := range stores {
+		wg.Add(1)
+		go func(addr string) {
```

Comment:
Is it possible to unify the goroutine usage from the `KVTxn.spawnWithStorePool` to make use of the pool?

Reply:
I don't get it. Did you mean using `c.store.Go()`?

Reply:
Yes, there are several wrappers based on the `kvstore.Go` interface, like the `KVTxn.spawnWithStorePool`. As a result there are different ways of spwaning goroutines in client-go by now, it's better to unify the usages if possible

Reply:
Gotcha. I didn't find `spawnWithStorePool` as it is recently merged